js.go

```diff
@@ -1250,6 +1335,127 @@ func (jsi *jsSub) trackSequences(reply string) {
 	jsi.mu.Unlock()
 }
 
+// Check to make sure messages are arriving in order.
+// Returns true if the sub had to be replaced. Will cause upper layers to return.
+// Lock should be held.
+func (sub *Subscription) checkOrderedMsgs(m *Msg) bool {
+	// Ignore msgs with no reply like HBs and flowcontrol, they are handled elsewhere.
+	if m.Reply == _EMPTY_ || sub.jsi == nil || isControlMessage(m) {
+		return false
+	}
+
+	// Normal message here.
+	tokens, err := getMetadataFields(m.Reply)
+	if err != nil {
+		return false
+	}
+	sseq, dseq := uint64(parseNum(tokens[5])), uint64(parseNum(tokens[6]))
+
+	jsi := sub.jsi
+	jsi.mu.Lock()
+	if dseq != jsi.dseq {
+		jsi.mu.Unlock()
+		sub.resetOrderedConsumer(jsi.sseq + 1)
+		return true
+	}
+	// Update our tracking here.
+	jsi.dseq, jsi.sseq = dseq+1, sseq
+	jsi.mu.Unlock()
+	return false
+}
+
+// We are here if we have detected a gap with an ordered consumer.
+// We will create a new consumer and rewire the low level subscription.
+// Lock should be held.
+func (sub *Subscription) resetOrderedConsumer(sseq uint64) {
+	nc := sub.conn
+	closed := sub.closed
+	if sub.jsi == nil || nc == nil || closed || nc.IsClosed() {
```

Comment:
Don't check nc.IsClosed() since this creates a lock inversion (allowed is CONN->SUB, not the other way around). Note that when connection is closed, the connection will mark all its sub.closed to true, so checking sub.closed should be enough here.

Reply:
Good point, will remove.